Userland/Applications/PDFViewer/ThumbnailsModel.cpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ThumbnailsModel.h"
+
+GUI::Variant ThumbnailsModel::data(GUI::ModelIndex const& index, GUI::ModelRole role) const
+{
+    if (role == GUI::ModelRole::TextAlignment)
+        return Gfx::TextAlignment::BottomCenter;
+    if (role != GUI::ModelRole::Display || !is_within_range(index))
+        return {};
+    return m_thumbnails.at(index.row());
+}
+
+void ThumbnailsModel::update_thumbnail(u32 index, NonnullRefPtr<Gfx::Bitmap const> thumbnail)
+{
+    if (index < m_thumbnails.size())
+        m_thumbnails[index] = thumbnail;
```

Comment:
It'd be nice if this was some kind of LRU thing. As is, once you scroll through all thumbnails, this will hold on to a thumbnail for every single page. At 128x128 pixels, a single thumbnail is 128x128x4 = 64Ki bytes. A 1000-page PDF will use 64Mi of ram just for the thumbnails.

(As a possible future improvement.)

Also, quickly scrolling through the thumbnail list will trigger 1000 page renders, all on the UI thread. It'd be nice to have have a system that has some more hysteresis, debouncing, and background processing – when an item becomes visible, it'd enqueue a "render thumbnail after 50ms" task, and when the item becomes invisible, it'd remove that task again if it's still not processed, etc.

It's a general list view thing almost – maybe the list view widget could grow support for lazily populated items. Maybe we could even use that for the ToC too then, to fix that other slow-loading big PDF issue I linked too.